File: src/util/ListUtil.fs
module ListUtil
    let tryMax cmp l =
        match l with
        | [] -> None
        | _ -> l |> List.sortWith cmp |> List.rev |> Some |> Option.map List.head

    let max cmp l =
        tryMax cmp l |> Option.get

    let tryMin cmp l =
        match l with
        | [] -> None
        | _ -> l |> List.sortWith cmp |> Some |> Option.map List.head

    let min cmp l =
        tryMin cmp l |> Option.get

    let makeList len v = List.init len (fun _ -> v)

    let tryLast l =
        match l with
        | [] -> None
        | _ -> l |> List.rev |> List.head |> Some

    let last l =
        tryLast l |> Option.get

    let rec take n l =
        match l with
        | [] -> []
        | _ :: _ when n <= 0 -> []
        | h :: t -> h :: take (n - 1) t

    let rec takeDrop n l =
        match l with
        | h :: t when n > 0 ->
            let l1, l2 = takeDrop (n - 1) t
            (h :: l1, l2)
        | l -> ([], l)

File: src/util/ListUtil.rs
//! Operations corresponding to the helpers in `ListUtil.fs`.

use std::cmp::Ordering;

pub fn try_max<T>(items: &[T], mut compare: impl FnMut(&T, &T) -> Ordering) -> Option<&T> {
    items.iter().max_by(|a, b| compare(a, b))
}

pub fn try_min<T>(items: &[T], mut compare: impl FnMut(&T, &T) -> Ordering) -> Option<&T> {
    items.iter().min_by(|a, b| compare(a, b))
}

pub fn make_list<T: Clone>(len: usize, value: T) -> Vec<T> { vec![value; len] }
pub fn try_last<T>(items: &[T]) -> Option<&T> { items.last() }

pub fn take<T: Clone>(count: isize, items: &[T]) -> Vec<T> {
    items.iter().take(count.max(0) as usize).cloned().collect()
}

pub fn take_drop<T: Clone>(count: isize, items: &[T]) -> (Vec<T>, Vec<T>) {
    let split = (count.max(0) as usize).min(items.len());
    (items[..split].to_vec(), items[split..].to_vec())
}
